File: android/app/src/main/java/io/github/junhezhang/qingjian/WebDavSync.java
package io.github.junhezhang.qingjian;

import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import android.util.Base64;

public final class WebDavSync {
    final String folder,authorization;
    static final class Response {int status;String etag;byte[] body;}
    public WebDavSync(SyncSettings s){this(s,false);}
    WebDavSync(SyncSettings s,boolean emulatorTest){if(!emulatorTest||!(s.url.startsWith("http://10.0.2.2:")||s.url.startsWith("http://127.0.0.1:")))s.validate();folder=s.url.replaceAll("/+$","")+"/";authorization="Basic "+Base64.encodeToString((s.user+":"+s.password).getBytes(StandardCharsets.UTF_8),Base64.NO_WRAP);}
    Response request(String method,String name,byte[] body,String match,boolean create)throws IOException {
        HttpURLConnection connection=(HttpURLConnection)new URL(folder+name).openConnection();connection.setRequestMethod(method);connection.setInstanceFollowRedirects(false);connection.setConnectTimeout(30000);connection.setReadTimeout(30000);connection.setUseCaches(false);connection.setRequestProperty("Authorization",authorization);connection.setRequestProperty("User-Agent","QingJian/0.2");
        if(match!=null)connection.setRequestProperty("If-Match",match);if(create)connection.setRequestProperty("If-None-Match","*");
        try{
            if(body!=null){connection.setDoOutput(true);connection.setRequestProperty("Content-Type","application/json; charset=utf-8");connection.setFixedLengthStreamingMode(body.length);try(OutputStream out=connection.getOutputStream()){out.write(body);}}
            Response r=new Response();r.status=connection.getResponseCode();r.etag=connection.getHeaderField("ETag");InputStream input=r.status>=400?connection.getErrorStream():connection.getInputStream();
            try(InputStream stream=input;ByteArrayOutputStream out=new ByteArrayOutputStream()){if(stream!=null){byte[] bytes=new byte[8192];int count;while((count=stream.read(bytes))!=-1){if(out.size()+count>SyncModel.MAX_BYTES)throw new IOException("服务器响应超过 4 MB。");out.write(bytes,0,count);}}r.body=out.toByteArray();}return r;
        }finally{connection.disconnect();}
    }
    static void require(Response r,int... values)throws IOException {for(int v:values)if(r.status==v)return;if(r.status==401||r.status==403)throw new IOException("网盘授权失败，请检查账号和应用密码（HTTP "+r.status+"）。");if(r.status==429)throw new IOException("网盘请求过于频繁，请稍后重试。");throw new IOException("同步服务器返回 HTTP "+r.status+"，本机待办未被替换。");}
    static String tag(Response r)throws IOException {String tag=r.etag;if(tag==null||tag.length()<2||!tag.startsWith("\"")||!tag.endsWith("\"")||tag.contains("\r")||tag.contains("\n"))throw new IOException("该 WebDAV 服务未提供强 ETag，无法安全处理同时修改。");return tag;}
    public void checkConnection()throws IOException {
        String name=".qingjian-probe-"+UUID.randomUUID()+".json";byte[] body="{\"probe\":1}".getBytes(StandardCharsets.UTF_8);require(request("PUT",name,body,null,true),200,201,204);
        try{requireConditionalRejection(request("PUT",name,body,null,true));requireConditionalRejection(request("PUT",name,body,"\"qingjian-must-not-match\"",false));Response read=request("GET",name,null,null,false);require(read,200);require(request("PUT",name,"{\"probe\":2}".getBytes(StandardCharsets.UTF_8),tag(read),false),200,201,204);}
        finally{require(request("DELETE",name,null,null,false),200,204,404);}
    }
    static void requireConditionalRejection(Response r)throws IOException {if(r.status>=200&&r.status<300)throw new IOException("该网盘未执行条件写入，不能安全同步。请联系服务商。");require(r,412);}
    static String digest(byte[] data)throws java.security.NoSuchAlgorithmException {return Base64.encodeToString(java.security.MessageDigest.getInstance("SHA-256").digest(data),Base64.URL_SAFE|Base64.NO_WRAP|Base64.NO_PADDING);}
    static void backup(File dir,String name,byte[] data)throws IOException {if(!dir.isDirectory()&&!dir.mkdirs())throw new IOException("无法创建同步备份目录。");File file=new File(dir,name);if(file.exists())return;try(FileOutputStream out=new FileOutputStream(file)){out.write(data);out.getFD().sync();}}
    public List<SyncModel.Item> exchange(List<SyncModel.Item> baseline,List<SyncModel.Item> local,File backupDir,SyncModel.Resolution resolution)throws Exception {
        checkConnection();
        for(int attempt=0;attempt<3;attempt++){
            Response read=request("GET","qingjian-tasks-v1.json",null,null,false);require(read,200,404);List<SyncModel.Item> remote=read.status==404?new ArrayList<>():SyncModel.decode(read.body);String etag=read.status==404?null:tag(read);
            if(!SyncModel.same(local,remote)){byte[] left=SyncModel.encode(local),right=SyncModel.encode(remote);String stamp=digest(left)+digest(right);backup(backupDir,stamp+"-local.json",left);backup(backupDir,stamp+"-remote.json",right);}
            List<SyncModel.Item> merged=SyncModel.merge(baseline,local,remote,resolution);if(read.status==200&&SyncModel.same(merged,remote))return merged;
            Response write=request("PUT","qingjian-tasks-v1.json",SyncModel.encode(merged),etag,etag==null);if(write.status==412)continue;require(write,200,201,204);return merged;
        }throw new IOException("另一台设备正在修改清单，本机修改已保留，请稍后重试。");
    }
}

File: WebDavSync.cs
using System;
using System.IO;
using System.Net;
using System.Text;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Json;
using System.Security.Cryptography;

namespace DesktopMemo
{
    [DataContract]
    public class SyncSettings
    {
        [DataMember] public string Url="https://dav.jianguoyun.com/dav/";
        [DataMember] public string User="";
        [DataMember] public string Password="";
        [DataMember] public bool Enabled;
        public string Binding() { using(var hash=SHA256.Create())return Convert.ToBase64String(hash.ComputeHash(Encoding.UTF8.GetBytes(Url.TrimEnd('/')+"\n"+User))); }
        public void Validate()
        {
            Uri uri;
            if(!Uri.TryCreate(Url,UriKind.Absolute,out uri)||uri.Scheme!="https"||uri.UserInfo!=""||uri.Query!=""||uri.Fragment!="")throw new InvalidDataException("服务器地址必须是 HTTPS 目录地址，不含账号、查询参数或片段。");
            if(string.IsNullOrWhiteSpace(User)||User.Contains(":")||User.Contains("\r")||User.Contains("\n")||string.IsNullOrEmpty(Password))throw new InvalidDataException("请填写网盘账号和应用密码。");
        }
        public static SyncSettings Load(string path)
        {
            if(!File.Exists(path))return new SyncSettings();
            byte[] clear=ProtectedData.Unprotect(File.ReadAllBytes(path),null,DataProtectionScope.CurrentUser);
            using(var input=new MemoryStream(clear))return (SyncSettings)new DataContractJsonSerializer(typeof(SyncSettings)).ReadObject(input);
        }
        public void Save(string path)
        {
            using(var output=new MemoryStream())
            {
                new DataContractJsonSerializer(typeof(SyncSettings)).WriteObject(output,this);
                byte[] encrypted=ProtectedData.Protect(output.ToArray(),null,DataProtectionScope.CurrentUser);
                Directory.CreateDirectory(Path.GetDirectoryName(path));string temp=path+".tmp";File.WriteAllBytes(temp,encrypted);
                if(File.Exists(path))File.Replace(temp,path,null);else File.Move(temp,path);
            }
        }
    }
    public sealed class WebDavSync
    {
        readonly Uri folder;
        readonly string authorization;
        public sealed class Response
        {
            public int Status; public string Etag; public byte[] Body;
        }
        public WebDavSync(SyncSettings settings) : this(settings,false) { }
        internal WebDavSync(SyncSettings settings,bool loopbackTest)
        {
            var uri=new Uri(settings.Url.TrimEnd('/')+"/");
            if(!loopbackTest || !uri.IsLoopback)settings.Validate();
            folder=uri;
            authorization="Basic "+Convert.ToBase64String(Encoding.UTF8.GetBytes(settings.User+":"+settings.Password));
        }
        Response Request(string method,string name,byte[] body=null,string match=null,bool create=false)
        {
            var request=(HttpWebRequest)WebRequest.Create(new Uri(folder,name));request.Method=method;request.AllowAutoRedirect=false;request.Timeout=30000;request.ReadWriteTimeout=30000;
            request.Headers[HttpRequestHeader.Authorization]=authorization;request.UserAgent="QingJian/1.3";
            request.CachePolicy=new System.Net.Cache.RequestCachePolicy(System.Net.Cache.RequestCacheLevel.NoCacheNoStore);
            if(match!=null)request.Headers[HttpRequestHeader.IfMatch]=match;
            if(create)request.Headers[HttpRequestHeader.IfNoneMatch]="*";
            if(body!=null){request.ContentType="application/json; charset=utf-8";request.ContentLength=body.Length;using(var stream=request.GetRequestStream())stream.Write(body,0,body.Length);}
            HttpWebResponse response;
            try{response=(HttpWebResponse)request.GetResponse();}
            catch(WebException ex){if(ex.Response==null)throw new IOException("无法连接同步服务器，请检查网络后重试。",ex);response=(HttpWebResponse)ex.Response;}
            using(response)
            using(var stream=response.GetResponseStream())
            using(var output=new MemoryStream())
            {
                var buffer=new byte[8192];int count;
                while((count=stream.Read(buffer,0,buffer.Length))>0){if(output.Length+count>SyncModel.MaxBytes)throw new InvalidDataException("服务器响应超过 4 MB，已停止同步。");output.Write(buffer,0,count);}
                return new Response {Status=(int)response.StatusCode,Etag=response.Headers[HttpResponseHeader.ETag],Body=output.ToArray()};
            }
        }
        static void Require(Response r,params int[] allowed)
        {
            if(Array.IndexOf(allowed,r.Status)>=0)return;
            if(r.Status==401||r.Status==403)throw new IOException("网盘授权失败，请检查账号和应用密码（HTTP "+r.Status+"）。");
            if(r.Status==429)throw new IOException("网盘请求过于频繁，请稍后同步。");
            throw new IOException("同步服务器返回 HTTP "+r.Status+"，本机待办未被替换。");
        }
        static string StrongTag(Response r)
        {
            string tag=r.Etag;
            if(string.IsNullOrEmpty(tag)||tag.Length<2||!tag.StartsWith("\"",StringComparison.Ordinal)||!tag.EndsWith("\"",StringComparison.Ordinal)||tag.Contains("\r")||tag.Contains("\n"))throw new IOException("该 WebDAV 服务未提供强 ETag，无法安全处理同时修改。请更换服务或联系服务商。");
            return tag;
        }
        public void CheckConnection()
        {
            string name=".qingjian-probe-"+Guid.NewGuid().ToString("N")+".json";byte[] body=Encoding.UTF8.GetBytes("{\"probe\":1}");
            Require(Request("PUT",name,body,null,true),200,201,204);
            try
            {
                RequireConditionalRejection(Request("PUT",name,body,null,true));
                RequireConditionalRejection(Request("PUT",name,body,"\"qingjian-must-not-match\""));
                var read=Request("GET",name);Require(read,200);string tag=StrongTag(read);
                Require(Request("PUT",name,Encoding.UTF8.GetBytes("{\"probe\":2}"),tag),200,201,204);
            }
            finally { var cleanup=Request("DELETE",name);Require(cleanup,200,204,404); }
        }
        static void RequireConditionalRejection(Response response)
        {
            if(response.Status>=200&&response.Status<300)throw new IOException("该网盘未执行条件写入，不能安全同步。请联系服务商。");
            Require(response,412);
        }
        static void Backup(string folder,SyncDocument local,SyncDocument remote)
        {
            Directory.CreateDirectory(folder);
            byte[] left=SyncModel.Encode(local),right=SyncModel.Encode(remote);
            string stamp;
            using(var hash=SHA256.Create())stamp=BitConverter.ToString(hash.ComputeHash(left)).Replace("-","")+BitConverter.ToString(hash.ComputeHash(right)).Replace("-","");
            string path=Path.Combine(folder,stamp);
            if(!File.Exists(path+"-local.json"))File.WriteAllBytes(path+"-local.json",left);
            if(!File.Exists(path+"-remote.json"))File.WriteAllBytes(path+"-remote.json",right);
        }
        public SyncDocument Exchange(SyncDocument baseline,SyncDocument local,string backupFolder,SyncResolution resolution=null)
        {
            CheckConnection();
            for(int attempt=0;attempt<3;attempt++)
            {
                var read=Request("GET","qingjian-tasks-v1.json");Require(read,200,404);
                var remote=read.Status==404?new SyncDocument():SyncModel.Decode(read.Body);
                string tag=read.Status==404?null:StrongTag(read);
                // Save both sides before merging, including conflicts, without credentials.
                if(!SyncModel.Equal(local,remote))
                {
                    Backup(backupFolder,local,remote);
                }
                var merged=SyncModel.Merge(baseline,local,remote,resolution);
                if(read.Status==200&&SyncModel.Equal(merged,remote))return merged;
                var write=Request("PUT","qingjian-tasks-v1.json",SyncModel.Encode(merged),tag,tag==null);
                if(write.Status==412)continue;
                Require(write,200,201,204);return merged;
            }
            throw new IOException("另一台设备正在修改云端清单，本机修改已保留，请稍后重试。");
        }
    }
}
